sql-plugin/src/main/scala/org/apache/spark/sql/rapids/complexTypeExtractors.scala

```diff
@@ -186,6 +186,99 @@ case class GpuGetMapValue(child: Expression, key: Expression)
+    (left.dataType, right.dataType) match {
+      case (arr: ArrayType, e2: IntegralType) if (e2 != LongType) =>
+        Seq(arr, IntegerType)
+      case (MapType(keyType, valueType, hasNull), e2) =>
+        TypeCoercion.findTightestCommonType(keyType, e2) match {
+          case Some(dt) => Seq(MapType(dt, valueType, hasNull), dt)
+          case _ => Seq.empty
+        }
+      case (l, r) => Seq.empty
+    }
+  }
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    (left.dataType, right.dataType) match {
+      case (_: ArrayType, e2) if e2 != IntegerType =>
+        TypeCheckResult.TypeCheckFailure(s"Input to function $prettyName should have " +
+          s"been ${ArrayType.simpleString} followed by a ${IntegerType.simpleString}, but it's " +
+          s"[${left.dataType.catalogString}, ${right.dataType.catalogString}].")
+      case (MapType(e1, _, _), e2) if (!e2.sameType(e1)) =>
+        TypeCheckResult.TypeCheckFailure(s"Input to function $prettyName should have " +
+          s"been ${MapType.simpleString} followed by a value of same key type, but it's " +
+          s"[${left.dataType.catalogString}, ${right.dataType.catalogString}].")
+      case (e1, _) if (!e1.isInstanceOf[MapType] && !e1.isInstanceOf[ArrayType]) =>
+        TypeCheckResult.TypeCheckFailure(s"The first argument to function $prettyName should " +
+          s"have been ${ArrayType.simpleString} or ${MapType.simpleString} type, but its " +
+          s"${left.dataType.catalogString} type.")
+      case _ => TypeCheckResult.TypeCheckSuccess
+    }
+  }
+
+  // Eventually we need something more full featured like
+  // GetArrayItemUtil.computeNullabilityFromArray
+  override def nullable: Boolean = true
+
+  override def doColumnar(lhs: GpuColumnVector, rhs: GpuColumnVector): ColumnVector =
+    throw new IllegalStateException("This is not supported yet")
+
+  override def doColumnar(lhs: Scalar, rhs: GpuColumnVector): ColumnVector =
+    throw new IllegalStateException("This is not supported yet")
+
+  override def doColumnar(lhs: GpuColumnVector, rhs: Scalar): ColumnVector = {
+    lhs.dataType match {
+      case _: ArrayType => {
+        if (rhs.isValid) {
+          if (rhs.getInt > 0) {
+            // SQL 1-based index
+            lhs.getBase.extractListElement(rhs.getInt - 1)
+          } else if (rhs.getInt == 0) {
+            throw new ArrayIndexOutOfBoundsException("SQL array indices start at 1")
+          } else {
+            lhs.getBase.extractListElement(rhs.getInt)
+          }
+        } else {
+          withResource(Scalar.fromNull(
+            GpuColumnVector.getNonNestedRapidsType(dataType))) { nullScalar =>
+            ColumnVector.fromScalar(nullScalar, lhs.getRowCount.toInt)
+          }
+        }
+      }
+      case _: MapType => {
+        lhs.getBase.getMapValue(rhs)
+      }
+    }
+  }
+
+  override def doColumnar(numRows: Int, lhs: Scalar, rhs: Scalar): ColumnVector =
```

Comment:
I am not sure, but what is the user case for this one ?

Reply:
This can happen in a few cases with optimizations. It is not common, but we need to cover it.  However this will not work until we can get scalar lists and scalar maps to work.  That is OK for now because it is very rare.